### DISCUSSION TITLE
Tips on translating to PyCUDA

### DISCUSSION THREAD

**Discussion by ceich** [pyopencl/pyopencl#545] [Q&A]
I have some image reconstruction code that uses PyOpenCL but I need to use PyCUDA (targeting Jetson AGX Xavier which doesn't support PyOpenCL). I have found a somewhat helpful [page](http://homepages.math.uic.edu/~jan/mcs572f16/mcs572notes/lec29.html), but could use a few more tips. Thanks!

**Comment by isuruf** [comment]
Slightly related. POCL with CUDA support works just fine on the Xavier and therefore you can keep using PyOpenCL.

**Reply by ceich** [reply]
Sorry, I'm a POCL n00b; how do I add CUDA support?

**Reply by ceich** [reply]
I began with this (3-year-old) [answer](https://forums.developer.nvidia.com/t/can-the-xavier-run-opencl-applications/70262) saying that POCL is _not_ supported on Xavier.

I (just this week) [asked what an alternative was and was told PyCUDA](https://forums.developer.nvidia.com/t/opencl-alternative/209242).

**Reply by isuruf** [reply]
There's an issue with using both CUDA device and CPU device as mentioned in https://github.com/pocl/pocl/issues/853,
but if you disable the CPU device, CUDA works. I don't have access to a Xavier, so I can't debug that.

**Reply by isuruf** [reply]
If you are using conda to install pyopencl (see https://documen.tician.de/pyopencl/misc.html#installing-from-conda-forge)
then you can use `conda install pocl-cuda` to get a pre-built pocl binary with CUDA support.
To avoid the CPU device, you can do `export POCL_DEVICES=cuda`

**Comment by inducer** [comment]
I agree with @isuruf that pocl should help offer OpenCL even where Nvidia does not. And with OpenCL being a much less stateful (i.e. much nicer!) API than CUDA, I think this is generally the technically preferable solution.

That said, if you must do a port, it's not too hard to come up with an abstraction layer, e.g.

- https://github.com/inducer/pyopencl/blob/main/pyopencl/_cluda.py
- https://reikna.publicfields.net/en/latest/introduction.html#cluda
